src/power_grid_model/core/power_grid_core.py

```diff
@@ -0,0 +1,220 @@
+        # call
+        res = self._cfunc(*c_inputs)
+        # convert to string for c_char_p
+        if self._c_restype == c_char_p:
+            res = res.decode()
+        return res
+
+
+class PowerGridCore:
+    """
+    DLL caller
+    """
+
+    _cdll: CDLL
+    _handle: HandlePtr
+    # error handling
+    error_code: Callable[[], int]
+    error_message: Callable[[], str]
+    n_failed_scenarios: Callable[[], int]
+    failed_scenarios: Callable[[], IdxPtr]  # type: ignore
+    batch_errors: Callable[[], CharDoublePtr]  # type: ignore
+    clear_error: Callable[[], None]
+    # batch
+    is_batch_independent: Callable[[], int]
+    is_batch_cache_topology: Callable[[], int]
+    # meta data
+    meta_n_datasets: Callable[[], int]
+    meta_dataset_name: Callable[[int], str]
+    meta_n_components: Callable[[str], int]
+    meta_component_name: Callable[[str, int], str]
+    meta_component_size: Callable[[str, str], int]
+    meta_component_alignment: Callable[[str, str], int]
+    meta_n_attributes: Callable[[str, str], int]
+    meta_attribute_name: Callable[[str, str, int], str]
+    meta_attribute_ctype: Callable[[str, str, str], str]
+    meta_attribute_offset: Callable[[str, str, str], int]
+    is_little_endian: Callable[[], int]
+    # options
+    create_options: Callable[[], OptionsPtr]
+    destroy_options: Callable[[OptionsPtr], None]
+    set_calculation_type: Callable[[OptionsPtr, int], None]
+    set_calculation_method: Callable[[OptionsPtr, int], None]
+    set_symmetric: Callable[[OptionsPtr, int], None]
+    set_err_tol: Callable[[OptionsPtr, float], None]
+    set_max_iter: Callable[[OptionsPtr, int], None]
+    set_threading: Callable[[OptionsPtr, int], None]
+    # model
+    create_model: Callable[[float, int, CharDoublePtr, IdxPtr, VoidDoublePtr], ModelPtr]  # type: ignore
+    update_model: Callable[[ModelPtr, int, CharDoublePtr, IdxPtr, VoidDoublePtr], None]  # type: ignore
+    copy_model: Callable[[ModelPtr], ModelPtr]
+    get_indexer: Callable[[ModelPtr, str, int, IDPtr, IdxPtr], None]  # type: ignore
+    calculate: Callable[  # type: ignore
+        [
+            # model
+            ModelPtr,
+            OptionsPtr,
+            # output
+            int,
+            CharDoublePtr,  # type: ignore
+            VoidDoublePtr,  # type: ignore
+            # update
+            int,
+            int,
+            CharDoublePtr,  # type: ignore
+            IdxPtr,
+            IdxDoublePtr,  # type: ignore
+            VoidDoublePtr,  # type: ignore
+        ],
+        None,
+    ]
+    destroy_model: Callable[[ModelPtr], None]
+
+    def __new__(cls, *args, **kwargs):
+        instance = super().__new__(cls, *args, **kwargs)
+        instance._cdll = _load_core()
+        instance._handle = instance._cdll.PGM_create_handle()
+        return instance
+
+    def __init__(self):
+        for name, function in PowerGridCore.__annotations__.items():  # pylint: disable=E1101
```

Comment:
Can we handle this with a decorator?

For example:
```python
@cbinding
def meta_dataset_name(self, dataset_idx: int) -> str:
    pass
```
or (my preference) something like:
```python
@cbinding
def meta_dataset_name(self, dataset_idx: int) -> str:
    return self._cdll.PGM_meta_dataset_name(self._handle, dataset_idx).decode()
```

Reply:
I.e. I would move the `argtypes` part from the for loop in `__init__` to the decorator function. And I would like to handle the type conversions (currently defined in WrapperFunc) explicitly in the function body.